e2etests/test/device_service.js

```diff
@@ -13,13 +13,55 @@ var createDeviceClient = require('./testUtils.js').createDeviceClient;
 var closeDeviceServiceClients = require('./testUtils.js').closeDeviceServiceClients;
 var closeDeviceEventHubClients = require('./testUtils.js').closeDeviceEventHubClients;
 var eventHubClient = require('azure-event-hubs').Client;
+var DeviceIdentityHelper = require('./device_identity_helper.js');
+
+var deviceAmqp = require('azure-iot-device-amqp');
+var deviceMqtt = require('azure-iot-device-mqtt');
+var deviceHttp = require('azure-iot-device-http');
+
+var hubConnectionString = process.env.IOTHUB_CONNECTION_STRING;
 
 var maximumMessageSize = ((256*1024)-512);
 
-var runTests = function (hubConnectionString, deviceTransport, provisionedDevice) {
-  describe('Device utilizing ' + provisionedDevice.authenticationDescription + ' authentication, connected over ' + deviceTransport.name + ' using device/service clients c2d', function () {
+[
+  DeviceIdentityHelper.createDeviceWithSas,
+  DeviceIdentityHelper.createDeviceWithSymmetricKey,
+  DeviceIdentityHelper.createDeviceWithX509SelfSignedCert
+].forEach(function (createDeviceMethod) {
+  [
+    deviceHttp.Http,
+    deviceAmqp.Amqp,
+    deviceAmqp.AmqpWs,
+    deviceMqtt.Mqtt
+  ].forEach(function (deviceTransport) {
+    device_service_tests(deviceTransport, createDeviceMethod);
+  });
+});
+
+[
+  deviceAmqp.Amqp,
+  deviceMqtt.Mqtt
+].forEach(function (deviceTransport) {
+  device_service_tests(deviceTransport, DeviceIdentityHelper.createDeviceWithX509CASignedCert);
+});
```

Comment:
kinda looks like you're doing this twice.

Reply:
it's a weird matrix.
in the first nested loops i do it for 3 authentication methods over 4 protocols, then I do it only over MQTT and AMQP for the CA signed cert authentication

---
In reply to: [149209475](https://github.com/Azure/azure-iot-sdk-node/pull/134#discussion_r149209475) [](ancestors = 149209475)